From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lumine1909 <133463833+Lumine1909@users.noreply.github.com>
Date: Tue, 9 Nov 2077 00:00:00 +0800
Subject: [PATCH] Leaves Fork cache item max stack size

Slightly modified and cleaned up version

Original license: GPL-3.0-only
Original project: https://github.com/Lumine1909/Leaves_Fork

diff --git a/net/minecraft/core/component/PatchedDataComponentMap.java b/net/minecraft/core/component/PatchedDataComponentMap.java
index 08ea8107fc8cbec3742278ab5cf03ef9048720f9..a13e8816c4aab403757a5488a5a3a8b9458c4c30 100644
--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -18,6 +18,7 @@ public final class PatchedDataComponentMap implements DataComponentMap {
     private final DataComponentMap prototype;
     private Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch;
     private boolean copyOnWrite;
+    private int maxStackSize; // Leaves - cache item max stack size
 
     public PatchedDataComponentMap(final DataComponentMap prototype) {
         this(prototype, Reference2ObjectMaps.emptyMap(), true);
@@ -29,6 +30,13 @@ public final class PatchedDataComponentMap implements DataComponentMap {
         this.prototype = prototype;
         this.patch = patch;
         this.copyOnWrite = copyOnWrite;
+        // Leaves start - cache item max stack size
+        if (!prototype.isEmpty() || !patch.isEmpty()) {
+            detectMaxStackSizeChange(null);
+        } else {
+            this.maxStackSize = 1;
+        }
+        // Leaves end - cache item max stack size
     }
 
     public static PatchedDataComponentMap fromPatch(final DataComponentMap prototype, final DataComponentPatch patch) {
@@ -76,6 +84,7 @@ public final class PatchedDataComponentMap implements DataComponentMap {
             lastValue = (Optional<T>)this.patch.put(type, Optional.ofNullable(value));
         }
 
+        this.detectMaxStackSizeChange(type); // Leaves - cache item max stack size
         return lastValue != null ? lastValue.orElse(defaultValue) : defaultValue;
     }
 
@@ -93,6 +102,7 @@ public final class PatchedDataComponentMap implements DataComponentMap {
             lastValue = (Optional<? extends T>)this.patch.remove(type);
         }
 
+        this.detectMaxStackSizeChange(type); // Leaves - cache item max stack size
         return (T)(lastValue != null ? lastValue.orElse(null) : defaultValue);
     }
 
@@ -117,17 +127,20 @@ public final class PatchedDataComponentMap implements DataComponentMap {
         } else {
             this.patch.remove(type);
         }
+        this.detectMaxStackSizeChange(type); // Leaves - cache item max stack size
     }
 
     public void restorePatch(final DataComponentPatch patch) {
         this.ensureMapOwnership();
         this.patch.clear();
         this.patch.putAll(patch.map);
+        this.detectMaxStackSizeChange(null); // Leaves - cache item max stack size
     }
 
     public void clearPatch() {
         this.ensureMapOwnership();
         this.patch.clear();
+        this.detectMaxStackSizeChange(null); // Leaves - cache item max stack size
     }
 
     public void setAll(final DataComponentMap components) {
@@ -233,4 +246,16 @@ public final class PatchedDataComponentMap implements DataComponentMap {
     public String toString() {
         return "{" + this.stream().map(TypedDataComponent::toString).collect(Collectors.joining(", ")) + "}";
     }
+
+    // Leaves start - cache item max stack size
+    public int getMaxStackSize() {
+        return this.maxStackSize;
+    }
+
+    private void detectMaxStackSizeChange(final @Nullable DataComponentType<?> changedType) {
+        if (changedType == null || changedType == DataComponents.MAX_STACK_SIZE) {
+            this.maxStackSize = this.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
+        }
+    }
+    // Leaves end - cache item max stack size
 }
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 59fdfa1e27b04d06de9912de5ed74ff4f86bf004..ffd9e54c62f86177f82d5d74d44c64368d126e23 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -1306,6 +1306,13 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
         return this.getCount();
     }
 
+    // Leaves start - cache item max stack size
+    @Override
+    public int getMaxStackSize() {
+        return this.components.getMaxStackSize();
+    }
+    // Leaves end - cache item max stack size
+
     public void setCount(final int count) {
         // Leaves start - Lithium Sleeping Block Entity
         if (count != this.count) {
